src/bin/ion/commands/beta/generate/model.rs

```diff
@@ -0,0 +1,268 @@
+use crate::commands::beta::generate::context::SequenceType;
+use serde::Serialize;
+
+/// Represent a node in the data model tree of the generated code.
+/// Each node in this tree could either be a module/package or a concrete data structure(class, struct, enum etc.).
+/// This tree structure will be used by code generator and templates to render the generated code as per given ISL type definition hierarchy.
+#[allow(dead_code)]
+#[derive(Debug, Clone, PartialEq)]
+pub struct DataModelNode {
+    // Represents the name of this data model
+    // Note: It doesn't point to the fully qualified name. To get fully qualified name use `fully_qualified_name()` from `AbstractDataType`.
+    name: String,
+    // Represents the type of the data model
+    // It can be `None` for modules or packages.
+    code_gen_type: Option<AbstractDataType>,
+    // Represents the nested types for this data model
+    nested_types: Vec<DataModelNode>,
+}
+
+impl DataModelNode {
+    #![allow(dead_code)]
+    pub fn is_scalar(&self) -> bool {
+        if let Some(code_gen_type) = &self.code_gen_type {
+            return matches!(code_gen_type, AbstractDataType::Scalar(_));
+        }
+        false
+    }
+
+    pub fn is_sequence(&self) -> bool {
+        if let Some(code_gen_type) = &self.code_gen_type {
+            return matches!(code_gen_type, AbstractDataType::Sequence(_));
+        }
+        false
+    }
+
+    pub fn is_structure(&self) -> bool {
+        if let Some(code_gen_type) = &self.code_gen_type {
+            return matches!(code_gen_type, AbstractDataType::Structure(_));
+        }
+        false
+    }
+}
+
+/// Represents a fully qualified type name for a type definition
+/// e.g. For a `Foo` class in `org.example` namespace
+///     In Java, `org.example.Foo`
+///     In Rust, `org::example::Foo`
+type FullyQualifiedTypeName = Vec<String>;
+
+/// Represents a fully qualified type name for a type reference
+#[allow(dead_code)]
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub struct FullyQualifiedTypeReference {
+    // Represents fully qualified name of the type
+    // e.g. In Java, `org.example.Foo`
+    //      In Rust, `crate::org::example::Foo`
+    type_name: FullyQualifiedTypeName,
+    // For types with parameters this will represent the nested parameter
+    parameters: Vec<FullyQualifiedTypeReference>,
+}
+
+/// A target-language-agnostic data type that determines which template(s) to use for code generation.
+#[allow(dead_code)]
+// TODO: Add more code gent types like sum/discriminated union, enum and map.
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub enum AbstractDataType {
+    // Represents a top level scalar type definition which also has a name attached to it.
+    WrappedScalar(WrappedScalar),
+    // Represents a nested/inline scalar value (e.g. a string or integer or user defined type)
+    Scalar(Scalar),
```

Comment:
The fundamental difference here is not how it is defined in ISL—it's whether the type is nominally a different type from its base type.